### PR TITLE
Using and testing non-default AWS credentials

### PR DESCRIPTION
**Pull request by metral**
### Proposed changes

The following changes are related to using and testing alternative AWS credentials, instead of the default credential chain, to create EKS clusters.

CI:
- Fixes CI smoke test bug
- Updates CI AWS auth to use `aws configure` to allow the use of alternative named profiles.

  _Note: TravisCI environment variables were also updated to work with `aws configure`._

Cluster:
- Pass an AWS provider to all AWS resources if one is set in the cluster resource options.

  _Note: Tested in CloudTrail for alternative profile use to `get`/`create` all AWS resources._

kubeconfig + non-default AWS creds:
- Adds a new cluster option: `providerCredentialOpts` to configure the cluster's kubeconfig auth when using a non-default AWS creds chain (e.g. such as using named profiles or AWS provider instance).
- Adds a new example/test for that uses a named profile and an AWS provider instance.

---

### Related issues (optional)

Fixes https://github.com/pulumi/pulumi-eks/issues/368
Fixes https://github.com/pulumi/pulumi-eks/issues/369
Closes https://github.com/pulumi/pulumi-eks/issues/370
Fixes https://github.com/pulumi/pulumi-eks/issues/296

### CODE REVIEW

File: nodejs/eks/cluster.ts [hunk 1]
```diff
@@ -144,15 +144,21 @@ export interface CoreData {
...
+    parent: pulumi.ComponentResource,
+    instanceRoleName?: pulumi.Input<aws.iam.Role>,
+    instanceProfileName?: pulumi.Input<string>,
+    provider?: pulumi.ProviderResource,
+): aws.iam.InstanceProfile {
     let instanceProfile: aws.iam.InstanceProfile;
     if (instanceProfileName) {
         instanceProfile = aws.iam.InstanceProfile.get(`${name}-instanceProfile`,
-            instanceProfileName, undefined, { parent: parent });
+            instanceProfileName, undefined, { parent: parent, provider });
```
Comment (lblackstone) [inline nodejs/eks/cluster.ts:158]: Should make arg usage consistent.
```suggestion
            instanceProfileName, undefined, { parent, provider });
```

File: nodejs/eks/cluster.ts [hunk 2]
```diff
@@ -144,15 +144,21 @@ export interface CoreData {
...
     } else {
         instanceProfile = new aws.iam.InstanceProfile(`${name}-instanceProfile`, {
             role: instanceRoleName,
-        }, { parent: parent });
+        }, { parent: parent, provider });
```
Comment (lblackstone) [inline nodejs/eks/cluster.ts:162]:
```suggestion
        }, { parent, provider });
```

File: nodejs/eks/cluster.ts [hunk 3]
```diff
@@ -364,7 +376,7 @@ export function createCore(name: string, args: ClusterOptions, parent: pulumi.Co
                 "arn:aws:iam::aws:policy/AmazonEKSClusterPolicy",
                 "arn:aws:iam::aws:policy/AmazonEKSServicePolicy",
             ],
-        }, { parent: parent })).role;
+        }, { parent: parent, provider })).role;
```
Comment (lblackstone) [inline nodejs/eks/cluster.ts:393]:
```suggestion
        }, { parent, provider })).role;
```

File: nodejs/eks/cluster.ts [hunk 4]
```diff
@@ -383,7 +395,7 @@ export function createCore(name: string, args: ClusterOptions, parent: pulumi.Co
                 ...clusterSecurityGroupTags,
                 ...tags,
             })),
-        }, { parent: parent });
+        }, { parent: parent, provider });
```
Comment (lblackstone) [inline nodejs/eks/cluster.ts:412]:
```suggestion
        }, { parent, provider });
```

File: nodejs/eks/cluster.ts [hunk 5]
```diff
@@ -393,7 +405,7 @@ export function createCore(name: string, args: ClusterOptions, parent: pulumi.Co
             protocol: "-1",  // all
             cidrBlocks: ["0.0.0.0/0"],
             securityGroupId: eksClusterSecurityGroup.id,
-        }, { parent: parent });
+        }, { parent: parent, provider });
```
Comment (lblackstone) [inline nodejs/eks/cluster.ts:422]:
```suggestion
        }, { parent, provider });
```

File: nodejs/eks/cluster.ts [hunk 6]
```diff
@@ -525,7 +537,7 @@ export function createCore(name: string, args: ClusterOptions, parent: pulumi.Co
                 "arn:aws:iam::aws:policy/AmazonEKS_CNI_Policy",
                 "arn:aws:iam::aws:policy/AmazonEC2ContainerRegistryReadOnly",
             ],
-        }, { parent: parent })).role;
+        }, { parent: parent, provider })).role;
```
Comment (lblackstone) [inline nodejs/eks/cluster.ts:556]:
```suggestion
        }, { parent, provider })).role;
```

File: nodejs/eks/cluster.ts [hunk 7]
```diff
@@ -596,7 +608,7 @@ export function createCore(name: string, args: ClusterOptions, parent: pulumi.Co
             managedPolicyArns: [
                 "arn:aws:iam::aws:policy/AmazonEKSFargatePodExecutionRolePolicy",
             ],
-        }, { parent: parent })).role.apply(r => r.arn);
+        }, { parent: parent, provider })).role.apply(r => r.arn);
```
Comment (lblackstone) [inline nodejs/eks/cluster.ts:627]:
```suggestion
        }, { parent, provider })).role.apply(r => r.arn);
```

File: nodejs/eks/cluster.ts [hunk 8]
```diff
@@ -608,7 +620,7 @@ export function createCore(name: string, args: ClusterOptions, parent: pulumi.Co
             podExecutionRoleArn: podExecutionRoleArn,
             selectors: selectors,
             subnetIds: pulumi.output(clusterSubnetIds).apply(subnets => computeWorkerSubnets(parent, subnets)),
-        }, { parent: parent, dependsOn: [eksNodeAccess] });
+        }, { parent: parent, dependsOn: [eksNodeAccess], provider});
```
Comment (lblackstone) [inline nodejs/eks/cluster.ts:639]:
```suggestion
        }, { parent, dependsOn: [eksNodeAccess], provider});
```

File: nodejs/eks/cluster.ts [hunk 9]
```diff
@@ -649,7 +661,7 @@ export function createCore(name: string, args: ClusterOptions, parent: pulumi.Co
             clientIdLists: ["sts.amazonaws.com"],
             url: eksCluster.identities[0].oidcs[0].issuer,
             thumbprintLists: [fingerprint],
-        }, { parent: parent });
+        }, { parent: parent, provider });
```
Comment (lblackstone) [inline nodejs/eks/cluster.ts:680]:
```suggestion
        }, { parent, provider });
```

File: nodejs/eks/nodegroup.ts [hunk 10]
```diff
@@ -358,7 +358,7 @@ export function createNodeGroup(name: string, args: NodeGroupOptions, parent: pu
     if (args.nodePublicKey) {
         const key = new aws.ec2.KeyPair(`${name}-keyPair`, {
             publicKey: args.nodePublicKey,
-        }, { parent: parent });
+        }, { parent: parent, provider });
```
Comment (lblackstone) [inline nodejs/eks/nodegroup.ts:361]:
```suggestion
        }, { parent, provider });
```

File: nodejs/eks/nodegroup.ts [hunk 11]
```diff
@@ -526,7 +526,7 @@ ${customUserData}
             ...cloudFormationTags,
             ...tags,
         })),
-    }, { parent: parent, dependsOn: cfnStackDeps });
+    }, { parent: parent, dependsOn: cfnStackDeps, provider });
```
Comment (lblackstone) [inline nodejs/eks/nodegroup.ts:529]:
```suggestion
    }, { parent, dependsOn: cfnStackDeps, provider });
```

File: nodejs/eks/nodegroup.ts [hunk 12]
```diff
@@ -783,7 +783,7 @@ export function createManagedNodeGroup(name: string, args: ManagedNodeGroupOptio
             };
         }),
         subnetIds: subnetIds,
-    }, { parent: parent ? parent : eksCluster, dependsOn: ngDeps });
+    }, { parent: parent ? parent : eksCluster, dependsOn: ngDeps, provider });
```
Comment (lblackstone) [inline nodejs/eks/nodegroup.ts:786]:
```suggestion
    }, { parent: parent ?? eksCluster, dependsOn: ngDeps, provider });
```

File: nodejs/eks/securitygroup.ts [hunk 13]
```diff
@@ -60,7 +60,7 @@ export function createNodeGroupSecurityGroup(name: string, args: NodeGroupSecuri
             [`kubernetes.io/cluster/${clusterName}`]: "owned",
             ...tags,
         })),
-    }, { parent: parent });
+    }, { parent: parent, provider });
```
Comment (lblackstone) [inline nodejs/eks/securitygroup.ts:63]:
```suggestion
    }, { parent, provider });
```

File: nodejs/eks/securitygroup.ts [hunk 14]
```diff
@@ -70,7 +70,7 @@ export function createNodeGroupSecurityGroup(name: string, args: NodeGroupSecuri
         protocol: "-1", // all
         securityGroupId: nodeSecurityGroup.id,
         self: true,
-    }, { parent: parent });
+    }, { parent: parent, provider });
```
Comment (lblackstone) [inline nodejs/eks/securitygroup.ts:73]:
```suggestion
    }, { parent, provider });
```

File: nodejs/eks/securitygroup.ts [hunk 15]
```diff
@@ -80,7 +80,7 @@ export function createNodeGroupSecurityGroup(name: string, args: NodeGroupSecuri
         protocol: "tcp",
         securityGroupId: nodeSecurityGroup.id,
         sourceSecurityGroupId: args.clusterSecurityGroup.id,
-    }, { parent: parent });
+    }, { parent: parent, provider });
```
Comment (lblackstone) [inline nodejs/eks/securitygroup.ts:83]:
```suggestion
    }, { parent, provider });
```

File: nodejs/eks/securitygroup.ts [hunk 16]
```diff
@@ -90,7 +90,7 @@ export function createNodeGroupSecurityGroup(name: string, args: NodeGroupSecuri
...
```
Comment (lblackstone) [inline nodejs/eks/securitygroup.ts:93]:
```suggestion
    }, { parent, provider });
```

File: nodejs/eks/securitygroup.ts [hunk 17]
```diff
@@ -100,7 +100,7 @@ export function createNodeGroupSecurityGroup(name: string, args: NodeGroupSecuri
         protocol: "-1",  // all
         cidrBlocks: [ "0.0.0.0/0" ],
         securityGroupId: nodeSecurityGroup.id,
-    }, { parent: parent });
+    }, { parent: parent, provider });
```
Comment (lblackstone) [inline nodejs/eks/securitygroup.ts:103]:
```suggestion
    }, { parent, provider });
```

File: nodejs/eks/securitygroup.ts [hunk 18]
```diff
@@ -110,7 +110,7 @@ export function createNodeGroupSecurityGroup(name: string, args: NodeGroupSecuri
         protocol: "tcp",
         securityGroupId: args.clusterSecurityGroup.id,
         sourceSecurityGroupId: nodeSecurityGroup.id,
-    }, { parent: parent });
+    }, { parent: parent, provider });
```
Comment (lblackstone) [inline nodejs/eks/securitygroup.ts:113]:
```suggestion
    }, { parent, provider });
```

File: nodejs/eks/examples/examples_test.go [hunk 19]
```diff
@@ -201,6 +200,25 @@ func TestAccRoleKubeconfig(t *testing.T) {
...
 }
 
+func TestAccAwsProfile(t *testing.T) {
+	// https://www.terraform.io/docs/providers/aws/index.html#shared-credentials-file
+	// https://docs.aws.amazon.com/sdk-for-go/api/aws/session/
+	// env := []string{fmt.Sprintf("%s=%d", "AWS_SDK_LOAD_CONFIG", 1)}
+	test := getJSBaseOptions(t).
+		With(integration.ProgramTestOptions{
+			Dir: path.Join(getCwd(t), "aws-profile"),
+			// Env: env,
```
Comment (lblackstone) [inline nodejs/eks/examples/examples_test.go:210]: Is this still needed?

**Comment by metral** [comment]
Feedback has been addressed. PTAL @lblackstone

File: nodejs/eks/cluster.ts [hunk 20]
```diff
@@ -1258,14 +1324,18 @@ export class Cluster extends pulumi.ComponentResource {
...
         return new NodeGroup(name, {
             ...args,
             cluster: this.core,
             nodeSecurityGroup: this.core.nodeGroupOptions.nodeSecurityGroup,
             clusterIngressRule: this.core.nodeGroupOptions.clusterIngressRule,
         }, {
             parent: this,
-            providers: { kubernetes: this.provider },
+            providers: {
+                ...awsProvider,
```
Comment (lblackstone) [inline nodejs/eks/cluster.ts:1336]: Does this work if `awsProvider` is `undefined`? I think the spread operator only applies to object types.

File: nodejs/eks/examples/aws-profile/package.json [hunk 21]
```diff
@@ -0,0 +1,10 @@
+{
+    "name": "example-aws-profile",
+    "devDependencies": {
+        "typescript": "^3.0.0"
+    },
+    "dependencies": {
+        "@pulumi/pulumi": "^1.0.0",
```
Comment (lblackstone) [inline nodejs/eks/examples/aws-profile/package.json:7]: Should we bump this to `2.0.0`, or do you want to hold off on that?

File: nodejs/eks/cluster.ts (hunk 20 above)
Comment (metral) [reply nodejs/eks/cluster.ts:1336]: Yes - https://www.bennadel.com/blog/3590-you-can-safely-apply-object-spread-to-null-and-undefined-values-in-typescript-3-2-4.htm

File: nodejs/eks/examples/aws-profile/package.json (hunk 21 above)
Comment (metral) [reply nodejs/eks/examples/aws-profile/package.json:7]: Good catch!

**Pull request merged**